drivers: regulator: add stm32 vrefbuf driver

tpambor opened the pull request:
This PR introduces a driver for the VREFBUF regulator available on STM32 microcontrollers. It also updates the device tree to include VREFBUF support for the STM32H5 and STM32U5 series.

Fixes #97249

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
Please use `LL_VREFBUF_xxx()` APIs from `#include <stm32_ll_system.h>`

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
```diff
@@ -0,0 +1,195 @@
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	config->base->CSR = config->base->CSR & ~VREFBUF_CSR_ENVR;
```
```suggestion
	LL_VREFBUF_Disable();
```

mathieuchopstm reviewed (commented):
Pass 1

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
```diff
@@ -0,0 +1,195 @@
+struct regulator_stm32_vrefbuf_voltage {
+	uint8_t vrs;
+	uint32_t uv;
+};
```
Step 1: change structure
```suggestion
struct regulator_stm32_vrefbuf_voltage {
	uint32_t vrs; /* LL_VREFBUF_VOLTAGE_SCALE_<X> */
	int32_t uv;   /* VREFBUF output in µV with this scale */
};
```

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
```diff
@@ -0,0 +1,195 @@
+	}
+
+	config->base->CSR =
+		(config->base->CSR & ~VREFBUF_CSR_VRS) | FIELD_PREP(VREFBUF_CSR_VRS, best_vrs);
```
Step 3: use LL functions
```suggestion
	LL_VREFBUF_SetVoltageScaling(vrs);
```

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
```diff
@@ -0,0 +1,195 @@
+	({		                                                                           \
+		.vrs = DT_PROP_BY_IDX(node_id, prop, UTIL_X2(idx)),		                   \
+		.uv = DT_PROP_BY_IDX(node_id, prop, UTIL_INC(UTIL_X2(idx)))	                   \
+	},), ())
```
Step 2: update binding and instantiation macro to consume only voltage
```suggestion
#define VREFBUF_VOLTAGE_ELEM(node_id, prop, idx)                                                   \
	({		                                                                           \
		.vrs = CONCAT(LL_VREFBUF_VOLTAGE_SCALE_, idx),		                   \
		.uv = DT_PROP_BY_IDX(node_id, prop, idx)	                   \
	},
```

mathieuchopstm commented on drivers/regulator/regulator_stm32_vrefbuf.c:
```diff
@@ -0,0 +1,187 @@
+	uint32_t best_vrs = 0;
+	int32_t best_voltage = INT32_MAX;
+
+	/* choose minimum possible voltage in range provided by the caller */
```
At least:
```suggestion
	/* search closest voltage above minimum requested by the caller */
```

But I'd suggest the following to remove ambiguity
```suggestion
	/*
	 * Search closest voltage above minimum requested by the caller
	 * among all voltages we support. Note that we must check all
	 * configurations, as the order is not consistent across series
	 * (i.e., whether VREFBUF0 < VREFBUF1 or VREFBUF0 > VREFBUF1).
	 */
```

mathieuchopstm reviewed (commented):
Besides DTSI values (and organization) I haven't checked, LGTM.

etienne-lms reviewed (commented):
LGTM. Values for H5 and U5 are consistent.
Minor comments.

erwango reviewed (commented):
Nitpick

tpambor commented:
@nordic-auko @seov-nordic Could you have a look?

tpambor committed:
tests: drivers: regulator: add test for STM32 vrefbuf regulator

Add build test for vrefbuf regulator found on STM32H5 and STM32U5 series.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
dts: arm: st: stm32u5: Add support for vrefbuf

Add a node for the VREFBUF peripheral in the STM32U5.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
dts: arm: st: stm32h5: Add support for vrefbuf

Add a node for the VREFBUF peripheral in the STM32H5.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
drivers: regulator: add stm32 vrefbuf driver

Add driver for the VREFBUF regulator found on STM32 microcontrollers.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
tests: drivers: regulator: move app.overlay to board-specific file

Relocate the contents of app.overlay into a board-specific overlay for
native_sim. This change prepares the codebase for adding regulator tests
for other boards.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

kartben commented:
@nordic-auko @seov-nordic please review or re-assign to @erwango as stm32 platform maintainer.
